http-client/src/main/java/io/micronaut/http/client/netty/DefaultHttpClient.java

```diff
@@ -1671,48 +1639,66 @@ private String getHostHeader(URI requestURI) {
         return host.toString();
     }
 
-    private <I> void prepareHttpHeaders(
-        ConnectionManager.PoolHandle poolHandle,
-        URI requestURI,
-        io.micronaut.http.HttpRequest<I> request,
-        HttpRequest nettyRequest,
-        boolean permitsBody) {
-        HttpHeaders headers = nettyRequest.headers();
-
-        if (!headers.contains(HttpHeaderNames.HOST)) {
-            headers.set(HttpHeaderNames.HOST, getHostHeader(requestURI));
-        }
-
-        if (!poolHandle.http2) {
-            if (poolHandle.canReturn()) {
-                headers.set(HttpHeaderNames.CONNECTION, HttpHeaderValues.KEEP_ALIVE);
-            } else {
-                headers.set(HttpHeaderNames.CONNECTION, HttpHeaderValues.CLOSE);
+    private NettyByteBody buildFormRequest(
+        MutableHttpRequest<?> request,
+        EventLoop eventLoop,
+        ThrowingFunction<HttpRequest, HttpPostRequestEncoder, HttpPostRequestEncoder.ErrorDataEncoderException> buildMethod
+    ) throws HttpPostRequestEncoder.ErrorDataEncoderException {
+        // this function acts like a wrapper around HttpPostRequestEncoder. HttpPostRequestEncoder
+        // takes a request + form data and transforms it to a request + bytes. Because we only want
+        // the bytes, we need to copy the data from the netty request back to the original
+        // MutableHttpRequest. This is just the Content-Type header, which sometimes gets an extra
+        // boundary specifier that we need.
+
+        // build the mock netty request (only the content-type matters)
+        HttpRequest nettyRequest = new DefaultHttpRequest(HttpVersion.HTTP_1_1, HttpMethod.GET, "/");
+        List<AsciiString> relevantHeaders = List.of(HttpHeaderNames.CONTENT_TYPE);
+        for (AsciiString header : relevantHeaders) {
```

Comment:
What is the point of defining a list of one item and then iterating it twice?

Reply:
Not sure. I don't want to touch this rn, it's from the old implementation (just moved). Maybe this happened as part of the form handling changes